runtime/parachains/src/scheduler.rs

```diff
@@ -707,57 +517,130 @@ impl<T: Config> Pallet<T> {
-						continue
-					}
-
-					let entry = ParathreadEntry {
-						claim: ParathreadClaim(core_assignment.para_id, collator),
-						retries: retries + 1,
-					};
-
-					if entry.retries <= config.parathread_retries {
-						queue.enqueue_entry(entry, config.parathread_cores);
-					}
+		let mut vec = vec![];
+		for core_assignment in Scheduled::<T>::take() {
+			vec.push(core_assignment);
+		}
+
+		//T::CoreAssigners::<T>::clear(&vec);
+	}
+
+	// on new session
+	fn reschedule_occupied_cores(cores: Vec<CoreOccupied>) {
+		for (core_idx, core) in cores.iter().enumerate() {
+			match core {
+				CoreOccupied::Free => continue,
+				CoreOccupied::Parachain => continue,
+				CoreOccupied::Parathread(entry) => T::AssignmentProvider::push_assignment_for_core(
+					CoreIndex(core_idx as u32),
+					Assignment::ParathreadA(entry.claim.clone()),
+				),
+			}
+		}
+	}
+
+	//
+	// Lookahead related functions
+	//
+	fn backing_lookahead() -> u32 {
+		<configuration::Pallet<T>>::config().scheduling_lookahead
+	}
+
+	fn read_lookahead(lookahead: &[Vec<Assignment>], core_idx: CoreIndex) -> Option<&Assignment> {
+		lookahead.get(core_idx.0 as usize).and_then(|la_vec| la_vec.first())
+	}
+
+	fn fill_lookahead() {
+		let n_lookahead = Self::backing_lookahead();
+		let n_session_cores = T::AssignmentProvider::session_core_count();
+		let mut push_back_buffer = Vec::with_capacity((n_session_cores * n_lookahead) as usize);
+
+		for core_idx in 0..n_session_cores {
+			for _ in 0..n_lookahead {
+				// TODO: try to fill lookahead while lookahead is not full OR pop_assignment() returns None
+				// doesn_t work! parachains assigner never returns None...
+				// TODO: write tests for optimal allocation; check invariants
+				match Self::add_to_lookahead(CoreIndex(core_idx)) {
+					Err((_, None)) => break,                           // TODO: logging
+					Err((_, Some(tup))) => push_back_buffer.push(tup), // TODO: logging
+					Ok(_) => (),
 				}
 			}
+		}
+
+		// We could not schedule these assignments due to conflicts
+		for (core_idx, ass) in push_back_buffer.into_iter().rev() {
+			T::AssignmentProvider::push_front_assignment_for_core(core_idx, ass);
+		}
+	}
+
+	fn add_to_lookahead(
+		core_idx: CoreIndex,
+	) -> Result<(), (String, Option<(CoreIndex, Assignment)>)> {
+		match T::AssignmentProvider::pop_assignment_for_core(core_idx) {
+			None => Err(("no assignments at core_idx".to_string(), None)),
+			Some(ass) => {
+				// TODO: remove core_idx dependency
+				let lookahead_info = ParaCoreMapping::<T>::get()
+					.get(&ass.para_id())
+					.unwrap_or_else(|| &LookaheadInfo { core_idx, n_in_lookahead: 0 });
+
+				if Lookahead::<T>::get().get(lookahead_info.core_idx.0 as usize).len() as u32 >=
+					Self::backing_lookahead()
+				{
+					// If assigner is parathreads and we don't pop, we'd be stuck trying to add
+					// the same assignment again and again until the top level loop is over.
```

Comment:
We really either should have an interface, where it does not matter to the scheduler whether the assignment is a parathread or not or we should have two separate interfaces to make this distinction visible.